### PR TITLE
cryptographic-payment-receipts

### LINKED ISSUE #16: Cryptographic payment receipts with verifiable proofs
## Summary

Generate cryptographic receipts for every successful API request, providing users with tamper-proof payment proofs that can be independently verified.

## Problem

Currently, after a successful payment verification, users receive an API response but **no proof of payment**. This creates issues:

| Problem | Impact |
|---------|--------|
| **No audit trail** | Users can't prove they paid for a service |
| **Disputes** | No evidence for chargeback/refund claims |
| **Compliance** | Missing records for accounting purposes |
| **Trust** | Users must trust the service without proof |

## Proposed Solution

Implement a receipt system where the server signs a structured receipt after each successful request.

### Receipt Flow

```mermaid
sequenceDiagram
    participant C as Client
    participant G as Gateway
    participant V as Verifier
    participant AI as OpenRouter

    C->>G: Request + Payment Signature
    G->>V: Verify Payment
    V-->>G: ✓ Valid
    G->>AI: Get Response
    AI-->>G: AI Result
    
    Note over G: Generate Receipt
    G->>G: Sign(receipt_data, server_key)
    
    G-->>C: Response + X-402-Receipt Header
```

### Receipt Structure

```json
{
  "receipt": {
    "id": "rcpt_a1b2c3d4e5f6",
    "version": "1.0",
    "timestamp": "2024-01-15T10:30:00Z",
    "payment": {
      "payer": "0x742d35Cc6634C0532925a3b844Bc9e7595f8fE21",
      "recipient": "0x2cAF48b4BA1C58721a85dFADa5aC01C2DFa62219",
      "amount": "0.001",
      "token": "USDC",
      "chain_id": 8453,
      "nonce": "9c311e31-eb30-420a-bced-c0d68bc89cea"
    },
    "service": {
      "endpoint": "/api/ai/summarize",
      "request_hash": "sha256:a1b2c3...",
      "response_hash": "sha256:d4e5f6..."
    }
  },
  "signature": "0x...",
  "server_public_key": "0x..."
}
```

### 1. Receipt Generation

Create `gateway/receipt.go`:

```go
type Receipt struct {
    ID        string         `json:"id"`
    Version   string         `json:"version"`
    Timestamp time.Time      `json:"timestamp"`
    Payment   PaymentDetails `json:"payment"`
    Service   ServiceDetails `json:"service"`
}

type PaymentDetails struct {
    Payer     string `json:"payer"`
    Recipient string `json:"recipient"`
    Amount    string `json:"amount"`
    Token     string `json:"token"`
    ChainID   int    `json:"chain_id"`
    Nonce     string `json:"nonce"`
}

type ServiceDetails struct {
    Endpoint     string `json:"endpoint"`
    RequestHash  string `json:"request_hash"`
    ResponseHash string `json:"response_hash"`
}

type SignedReceipt struct {
    Receipt         Receipt `json:"receipt"`
    Signature       string  `json:"signature"`
    ServerPublicKey string  `json:"server_public_key"`
}

func GenerateReceipt(payment PaymentContext, payer string, 
                     endpoint string, reqBody, respBody []byte) (*SignedReceipt, error) {
    receipt := Receipt{
        ID:        generateReceiptID(),
        Version:   "1.0",
        Timestamp: time.Now().UTC(),
        Payment: PaymentDetails{
            Payer:     payer,
            Recipient: payment.Recipient,
            Amount:    payment.Amount,
            Token:     payment.Token,
            ChainID:   payment.ChainID,
            Nonce:     payment.Nonce,
        },
        Service: ServiceDetails{
            Endpoint:     endpoint,
            RequestHash:  hashData(reqBody),
            ResponseHash: hashData(respBody),
        },
    }
    
    return signReceipt(receipt)
}
```

### 2. EIP-712 Typed Signing (Server-side)

```go
func signReceipt(receipt Receipt) (*SignedReceipt, error) {
    // Serialize receipt deterministically
    receiptBytes, _ := json.Marshal(receipt)
    hash := crypto.Keccak256Hash(receiptBytes)
    
    // Sign with server's private key
    privateKey := getServerPrivateKey()
    signature, err := crypto.Sign(hash.Bytes(), privateKey)
    if err != nil {
        return nil, err
    }
    
    return &SignedReceipt{
        Receipt:         receipt,
        Signature:       hexutil.Encode(signature),
        ServerPublicKey: getServerPublicKey(),
    }, nil
}
```

### 3. Receipt Verification (Client-side)

```typescript
function verifyReceipt(signedReceipt: SignedReceipt): boolean {
  const receiptBytes = JSON.stringify(signedReceipt.receipt);
  const hash = ethers.keccak256(ethers.toUtf8Bytes(receiptBytes));
  
  const recoveredAddress = ethers.verifyMessage(
    ethers.getBytes(hash),
    signedReceipt.signature
  );
  
  // Verify signature matches known server public key
  const serverAddress = ethers.computeAddress(signedReceipt.serverPublicKey);
  return recoveredAddress.toLowerCase() === serverAddress.toLowerCase();
}
```

### 4. Response Integration

Update response to include receipt:

```go
func handleSummarize(c *gin.Context) {
    // ... existing verification and AI call ...
    
    // Generate receipt
    receipt, err := GenerateReceipt(
        context, 
        verifyResp.RecoveredAddress,
        c.Request.URL.Path,
        requestBody,
        responseBody,
    )
    
    // Add receipt to response header (base64 encoded)
    c.Header("X-402-Receipt", encodeReceipt(receipt))
    
    c.JSON(200, gin.H{
        "result":  summary,
        "receipt": receipt,  // Also include in body
    })
}
```

### 5. Receipt Lookup Endpoint

```
GET /api/receipts/:receipt_id

Response:
{
  "receipt": { ... },
  "signature": "0x...",
  "status": "valid"
}
```

## Architecture

```mermaid
flowchart TB
    subgraph Client
        APP[Application]
        VERIFY[Receipt Verifier]
    end
    
    subgraph Gateway
        HANDLER[Request Handler]
        RECEIPT[Receipt Generator]
        SIGNER[Server Signer]
        STORE[(Receipt Store)]
    end
    
    APP -->|Request| HANDLER
    HANDLER -->|Success| RECEIPT
    RECEIPT --> SIGNER
    SIGNER -->|Sign| RECEIPT
    RECEIPT -->|Store| STORE
    RECEIPT -->|Return| APP
    
    APP -->|Verify Locally| VERIFY
    APP -->|Lookup| STORE
```

## Configuration

```env
# Receipt Configuration
RECEIPT_ENABLED=true
RECEIPT_SIGNING_KEY=${SERVER_WALLET_PRIVATE_KEY}
RECEIPT_STORE_TYPE=memory    # memory | redis | postgres
RECEIPT_TTL_DAYS=365
```

### Core Features
- [ ] Create `Receipt` and `SignedReceipt` structs
- [ ] Implement deterministic JSON serialization
- [ ] Hash request/response bodies using SHA-256
- [ ] Sign receipts using server's private key
- [ ] Generate unique receipt IDs (`rcpt_` prefix + random)
- [ ] Include receipt in response header (`X-402-Receipt`)
- [ ] Include receipt in response body

### Verification
- [ ] Provide TypeScript/JavaScript verification library
- [ ] Add verification example to documentation
- [ ] Implement `GET /api/receipts/:id` lookup endpoint

### Storage
- [ ] Implement in-memory receipt store (development)
- [ ] Add optional Redis/PostgreSQL persistence
- [ ] Automatic cleanup of expired receipts

### Security
- [ ] Use constant-time comparison for signatures
- [ ] Rate limit receipt lookups
- [ ] Do NOT include full request/response in receipt (only hashes)
- [ ] Validate receipt format before storage

### Testing
- [ ] Unit tests for receipt generation
- [ ] Unit tests for signature verification
- [ ] Integration tests for full flow
- [ ] Test receipt verification with different clients

## Example Usage

**Response with receipt:**
```json
{
  "result": "This is the AI-generated summary...",
  "receipt": {
    "id": "rcpt_x7y8z9",
    "version": "1.0",
    "timestamp": "2024-01-15T10:30:00Z",
    "payment": {
      "payer": "0x742d...",
      "amount": "0.001",
      "nonce": "abc-123"
    },
    "service": {
      "endpoint": "/api/ai/summarize",
      "request_hash": "sha256:a1b2c3...",
      "response_hash": "sha256:d4e5f6..."
    },
    "signature": "0x...",
    "server_public_key": "0x..."
  }
}
```

**Client verification:**
```typescript
const response = await fetch('/api/ai/summarize', { ... });
const data = await response.json();

const isValid = verifyReceipt(data.receipt);
console.log(`Receipt valid: ${isValid}`);

// Store receipt for future reference
await localStorage.setItem(`receipt_${data.receipt.id}`, JSON.stringify(data.receipt));
```

## Resources

- [EIP-712 Typed Signing](https://eips.ethereum.org/EIPS/eip-712)
- [go-ethereum crypto package](https://pkg.go.dev/github.com/ethereum/go-ethereum/crypto)
- [JSON Canonicalization](https://datatracker.ietf.org/doc/html/rfc8785)

## PR Requirements

> ⚠️ **All PRs must include tests.** PRs without adequate test coverage will not be merged.

- [ ] Unit tests for receipt generation and ID uniqueness
- [ ] Unit tests for cryptographic signing
- [ ] Unit tests for signature verification
- [ ] Integration test for full receipt flow (generate → store → retrieve)
- [ ] TypeScript test for client-side verification
- [ ] Tests must pass: `go test -v ./...` and `bun test`

### PR DESCRIPTION
**Pull request by vishal0316**
<h2>Greptile Overview</h2>
closes #16 
### Greptile Summary

This PR implements cryptographic payment receipts for the MicroAI-Paygate system, providing tamper-proof proof-of-payment for every successful API request.

**Major changes:**
- Added receipt generation with ECDSA signatures (Keccak256) in `gateway/receipt.go`
- Integrated receipt creation into the payment flow after AI responses
- Added `GET /api/receipts/:id` endpoint for receipt lookup
- Implemented in-memory receipt storage with TTL-based cleanup
- Added TypeScript client library for receipt verification in `web/src/lib/verify-receipt.ts`
- Comprehensive test coverage (10 tests including integration tests)

**Issues found:**
- **JSON serialization mismatch**: The TypeScript verification library uses `JSON.stringify` which doesn't guarantee field order matches Go's `json.Marshal`. This could cause signature verification to fail if field orders diverge.
- **Misleading comment**: The comment in `receipt.go` claims `json.Marshal` serializes "alphabetically by JSON tag", but Go actually serializes in struct field definition order.
- **Request body reading**: Body is read twice (parsing + hashing), which could allow large invalid payloads to consume memory before validation.
- **Key validation**: Accepts keys as short as 16 bytes with left-padding, which may mask truncated/invalid configurations.

**Positive aspects:**
- Well-structured architecture with `ReceiptStore` interface for future extensibility
- Comprehensive test suite covering edge cases and TTL behavior
- Good security practices (constant-time operations, SHA-256 hashing, proper random ID generation)
- Thread-safe storage with proper mutex usage
- Proper cleanup goroutine with graceful shutdown handling
- 10MB body size limit to prevent DoS attacks

### Confidence Score: 4/5

- This PR is generally safe to merge with minor issues that should be monitored in testing
- Score reflects a solid implementation with good security practices and comprehensive tests. The main concern is the potential JSON serialization mismatch between Go and TypeScript which could affect receipt verification, though this is mitigated by Go's deterministic struct field ordering. The code includes proper DoS protections, thread-safe storage, and thorough test coverage. Minor issues like misleading comments and weak key validation are non-critical.
- Monitor `web/src/lib/verify-receipt.ts` in testing to ensure JSON serialization order matches Go's behavior for signature verification

<h3>Important Files Changed</h3>

File Analysis

| Filename | Score | Overview |
|----------|-------|----------|
| web/src/lib/verify-receipt.ts | 4/5 | TypeScript receipt verification library using ethers.js for signature recovery. Implementation looks correct with proper recovery ID adjustment (v+27). |
| gateway/receipt.go | 4/5 | Receipt generation and signing implementation using Keccak256 and ECDSA. Clean code with good security practices and proper error handling. |
| gateway/receipt_test.go | 5/5 | Comprehensive test suite with 10 tests covering unit tests, integration tests, and TTL behavior. Well-structured and thorough. |
| gateway/main.go | 4/5 | Receipt integration into handleSummarize with proper storage, cleanup goroutine, and GET endpoint. Includes body size limits and proper error handling. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant C as Client
    participant G as Gateway
    participant V as Verifier
    participant AI as OpenRouter
    participant S as Receipt Store

    Note over C,G: Payment verification (existing flow)
    C->>G: POST /api/ai/summarize + X-402-Signature
    G->>V: Verify signature
    V-->>G: Valid ✓ + recovered_address
    
    Note over G,AI: AI request processing
    G->>AI: Forward summarize request
    AI-->>G: AI response

    Note over G,S: NEW: Receipt generation & storage
    G->>G: Hash request body (SHA-256)
    G->>G: Hash response body (SHA-256)
    G->>G: Create Receipt struct
    G->>G: Sign with server private key (ECDSA)
    G->>S: Store receipt with 24h TTL
    S-->>G: Stored ✓
    
    Note over G,C: Response with receipt
    G-->>C: 200 OK + result + receipt (JSON & header)
    
    Note over C: Client-side verification
    C->>C: Verify receipt signature
    C->>C: Store receipt locally
    
    Note over C,G: Receipt lookup (new endpoint)
    C->>G: GET /api/receipts/rcpt_abc123
    G->>S: Retrieve by ID
    S-->>G: Receipt data or 404
    G-->>C: Receipt with signature + status
```

### CODE REVIEW

File: gateway/main.go [hunk 1]
```diff
@@ -527,3 +577,148 @@ func getEnvAsInt(key string, defaultValue int) int {
...
+		return
+	}
+
+	c.JSON(200, gin.H{
+		"receipt":           receipt.Receipt,
+		"signature":         receipt.Signature,
+		"server_public_key": receipt.ServerPublicKey,
+		"status":            "valid",
+	})
+}
```
Comment (AnkanMisra) [inline gateway/main.go]: `handleGetReceipt` endpoint has no authentication or rate limiting any attackers could brute-force receipt IDs 2^48 possibilities

apply a rate limit middleware to this endpoint

File: gateway/main.go [hunk 2]
```diff
@@ -269,7 +293,33 @@ func handleSummarize(c *gin.Context) {
...
+	receipt, err := GenerateReceipt(paymentCtx, verifyResp.RecoveredAddress, c.Request.URL.Path, requestBody, responseBody)
+	if err != nil {
+		log.Printf("error generating receipt: %v", err)
+		c.JSON(500, gin.H{"error": "Failed to generate receipt", "details": err.Error()})
+		return
+	}
+
+	// 6. Store receipt with TTL
+	storeReceipt(receipt, getReceiptTTL())
+
```
Comment (AnkanMisra) [inline gateway/main.go]: `storeReceipt` doesn't return error but the `ReceiptStore` interface declares `Store(...) error`  

Future Redis/Postgres implementations need this

**Review by AnkanMisra — CHANGES_REQUESTED** [review]
@vishal0316 

### Code Issues to Fix
1. Fix all the vulnerabilities that i find while reveiwng the code like 8 comments

And You missed some requirment that were mentioned in issue #16 

**Required:**
- [ ] TypeScript/JavaScript verification library
- [ ] Verification example in documentation
- [ ] Constant-time signature comparison (security)
- [ ] Receipt format validation before storage
- [ ] Full integration test with client verification

**Optional (separate PR ok):**
- [ ] Redis/PostgreSQL persistence


Good implementation of core features btw!
This is a hard issue, take a day or two no body is forcing

File: gateway/main.go [hunk 3]
```diff
@@ -191,6 +205,20 @@ func handleSummarize(c *gin.Context) {
...
+	// Capture request body for receipt generation
+	// Limit request body to 10MB to prevent memory exhaustion attacks
+	maxBodySize := int64(10 * 1024 * 1024)
+	c.Request.Body = http.MaxBytesReader(c.Writer, c.Request.Body, maxBodySize)
+
+	requestBody, err := c.GetRawData()
+	if err != nil {
+		log.Printf("error reading request body: %v", err)
+		c.JSON(500, gin.H{"error": "Failed to read request body"})
+		return
```
Comment (AnkanMisra) [inline gateway/main.go]: `MaxBytesReader` returns error when body > 10MB.

HTTP standard says this should return `413 Payload Too Large` , not `500 Internal Server Error`

Return 413, but not 500

File: gateway/main.go [hunk 4]
```diff
@@ -191,6 +205,20 @@ func handleSummarize(c *gin.Context) {
...
+	}
+	// Restore body for BindJSON later
+	c.Request.Body = http.NoBody // We'll unmarshal from requestBody directly
```
Comment (AnkanMisra) [inline gateway/main.go]: Misleading 
"BindJSON" but line 284 uses `json.Unmarshal(requestBody, &req)`

Update or remove

**Review by AnkanMisra — CHANGES_REQUESTED** [review]
@vishal0316 5 code issues to fix

Your Pr still missing the following feature that were mentioned in issue:

1. TypeScript/JavaScript verification library 
2. Verification example in documentation 
3. Constant-time signature comparison 
4. Receipt format validation before storage
5. Redis/PostgreSQL persistence (optional)

**Comment by vishal0316** [comment]
# Requirements Status 
@AnkanMisra 

## 1.  Receipt Format Validation
- Comprehensive `validateReceipt()` function (**lines 664–729**)
- Checks all required fields before storage
- Invoked inside `storeReceipt()` (**line 644**)

## 2.  Constant-Time Signature Comparison
- Documented in `receipt.go` (**lines 114–116**)
- Documented in `receipt_test.go` (**line 281**)
- Uses **go-ethereum** constant-time cryptographic operations

## 3.  TypeScript Verification Library
- **File:** `web/src/lib/verify-receipt.ts`
- **Functions:**
  - `verifyReceipt()`
  - `validateReceiptFormat()`
  - `fetchReceipt()`
- Full TypeScript type definitions included
- Usage examples provided

## 4.  Verification Documentation
- **README section:** *Receipt Verification* (**lines 389–524**)
- Includes:
  - TypeScript examples
  - API documentation
  - Security features explanation
  - Configuration guide

## 5. ⚠️ Redis / PostgreSQL Persistence (Optional)
- Marked as **OPTIONAL** for a future PR 
i will create new PR And solve this

**Comment by AnkanMisra** [comment]
@greptile review and check for TypeScript verification library and Integration test where to add (TestReceiptFullFlowIntegration)

**Comment by vishal0316** [comment]
> @greptile review and check for TypeScript verification library and Integration test where to add (TestReceiptFullFlowIntegration)

i fixed CODE REVIEW ISSUES

**Review by AnkanMisra — APPROVED** [review]
